Analysis/src/TypeInfer.cpp

```diff
@@ -1701,61 +1710,174 @@ void TypeChecker::check(const ScopePtr& scope, const AstStatDeclareClass& declar
 
     TypeId classTy = addType(ClassTypeVar(className, {}, superTy, std::nullopt, {}, {}, currentModuleName));
     ClassTypeVar* ctv = getMutable<ClassTypeVar>(classTy);
-
     TypeId metaTy = addType(TableTypeVar{TableState::Sealed, scope->level});
-    TableTypeVar* metatable = getMutable<TableTypeVar>(metaTy);
 
     ctv->metatable = metaTy;
-
     scope->exportedTypeBindings[className] = TypeFun{{}, classTy};
+}
 
-    for (const AstDeclaredClassProp& prop : declaredClass.props)
+void TypeChecker::check(const ScopePtr& scope, const AstStatDeclareClass& declaredClass)
+{
+    if (FFlag::LuauDeclareClassPrototype)
     {
-        Name propName(prop.name.value);
-        TypeId propTy = resolveType(scope, *prop.ty);
+        Name className(declaredClass.name.value);
+
+        // Don't bother checking if the class definition was incorrect
+        if (incorrectClassDefinitions.find(&declaredClass))
+            return;
+
+        std::optional<TypeFun> binding;
+        if (auto it = scope->exportedTypeBindings.find(className); it != scope->exportedTypeBindings.end())
+            binding = it->second;
+
+        // This class definition must have been `prototype()`d first.
+        if (!binding)
+            ice("Not predeclared");
```

Comment:
```suggestion
            ice("Class not predeclared");
```
A different message will make it easier to disambiguate with the one for type aliases.